docs/proposals/kubelet-eviction.md

```diff
@@ -0,0 +1,200 @@
+
+## Eviction scenario
+
+Let's assume the operator started the `kubelet` with the following:
+
+```
+--eviction-hard="memory.available<100Mi"
+--eviction-soft="memory.available<300Mi"
+--eviction-soft-period="memory.available=30s"
+```
+
+The `kubelet` will run a sync loop that looks at the available memory
+on the node as reported from `cAdvisor` by calculating (capacity - workingSet).
+If available memory is observed to drop below 100Mi, the `kubelet` will immediately
+initiate eviction. If available memory is observed as falling below `300Mi`,
+it will record when that signal was observed internally in a cache.  If at the next
+sync, that criteria was no longer satisfied, the cache is cleared for that
+signal.  If that signal is observed as being satisfied for longer than the
+specified period, the `kubelet` will initiate eviction to attempt to
+reclaim the resource that has met its eviction threshold.
+
+## Eviction of Pods
+
+If an eviction threshold has been met, the `kubelet` will initiate the
+process of evicting pods until it has observed the signal has gone below
+its defined threshold.
+
+The eviction sequence works as follows:
+
+* for each monitoring interval, if eviction thresholds have been exceeded
+ * find candidate pod
+ * fail the pod
+ * wait until pod is terminated
+
+This means that there is no parallel killing of pods, and we will wait
+a single monitoring interval before killing an additional pod to allow
+for reclamation of the starved resource to take place.  By default, the
+`kubelet` will respect graceful termination period when doing eviction.
+
+The mechanism of finding a candidate pod for eviction is tied to an
+eviction strategy.  An eviction strategy is a compiled in plug-in that
+delegates finding a candidate pod.  The `kubelet` is configured to work
+against a particular strategy.  The default strategy if not specified is
+`greedy`.  Other strategies are anticipated in the future.
+
+```
+--eviction-strategy="greedy": Strategy to use when eviction thresholds have been met.
+```
+
+### Eviction Strategy: greedy
+
+The `greedy` strategy targets pods that are the largest consumers of the
+starved compute resource relative to their scheduling guarantee.
+
+It ranks pods within a quality of service tier in the following priority
+order:
+
+* `BestEffort` pods that consume the most of the starved resource are failed
+first
+* `Burstable` pods that consume the greatest amount of the starved resource
+relative to their request for that resource are killed first.  If no pod
+has exceeded its request, the strategy targets the largest consumer of the
+starved resource.
+* `Guaranteed` pods that consume the greatest amount of the starved resource
+relative to their request are killed first.  If no pod has exceeded its request,
+the strategy targets the largest consumer of the starved resource.
+
+### Alternative eviction strategies
+
+It is anticipated other eviction strategies may be used in the future.
```

Comment:
These strategies can be added to the default `greedy` policy as well, since it is generally beneficial to all users.